README
Keyword Ranking

Use this gem to calculate the keyword ranking for Bing, Yahoo or Google as follows:

KeywordRanking::Ranking.get(:keyword => 'one_keyword', :url => 'www.mydomain.com', :engine => :bing, :limit => 100)

Note:

* :engine can receive any of these values -->  :google, :yahoo or :bing
* :limit by default is 200 (defined in RES_LIMIT constant)
* You can change the results per page desired changing the value of RES_PER_PAGE having in mind that not all the 3 engines support more than 10 results per page with this utility

More updates coming

TODO:

* Testing code



CHANGELOG
v1.0. First draft


spec/spec_helper.rb
$LOAD_PATH.unshift(File.dirname(__FILE__))
$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
require 'keyword_ranking'
require 'spec'
require 'spec/autorun'

Spec::Runner.configure do |config|
  
end


spec/keyword_ranking_spec.rb
require File.expand_path(File.dirname(__FILE__) + '/spec_helper')

# describe "KeywordRanking" do
#   it "fails" do
#     fail "hey buddy, you should probably rename this file and start specing for real"
#   end
# end
# 
# 
# # $LOAD_PATH << File.expand_path('../../lib' , __FILE__)
# # 
# # require 'rubygems'
# # require 'bundler'
# # Bundler.setup(:development)
# # require 'keyword_ranking'

module FinderSpecHelper
  class Finder
    def find(keyword, site, limit, engine)
      0
    end
  end
end

module KeywordRanking

    describe Ranking do

      let (:options) { {:keyword => 'agile', :url => 'www.crowdint.com', :engine => 'google', :limit => 20} }
      let(:ranking) { Ranking.new(FinderSpecHelper::Finder.new) }

      context '#get ranking fails' do

        it 'without parameters' do
          lambda { ranking.get }.should raise_exception
        end

        it 'with unsupported engine' do
          lambda { ranking.get(options.merge({:engine => 'mugle'})) }.should raise_exception
        end 

        it 'exceeding results limit' do
          lambda { ranking.get(options.merge({:limit => 201})) }.should raise_exception
        end 

        it 'with bad keyword and url data types' do
          lambda { ranking.get(options.merge({:keyword  => 1000, :url => 1.2})) }.should raise_exception
        end 

        it 'using malformed URLs'

      end 

      context '#get ranking success' do

        it 'with correct parameters on google' do
          ranking.get(options).should equal 0
        end

        it 'with correct parameters on yahoo'

        it 'with correct parameters on bing'

        it 'should support multiple keywords'

        it 'should support a symbol as an engine name'

      end

    end 

end


lib/keyword_ranking.rb
require 'rubygems'
require 'nokogiri'
require 'net/http'

class Array
  def extract_options!
    last.is_a?(::Hash) ? pop : {}
  end
end

class Ranking
  @@default_options = { :limit => 200, :res_per_page => 10, :supported_engines => [:bing, :google, :yahoo] } #careful, res_per_page doesn't work with more than 10 for all search engines!

  attr_reader :options

  def initialize(*options) # keyword, url, limit
    self.options = options.extract_options!
  end

  def options=(options)
    @options = @@default_options.merge(options)
    @options[:url].gsub!(/(http:\/\/|https:\/\/)/, '')
    @options[:url].gsub!(/^www/, '')
  end

  def from
    validate_options
    @finder ||= Finder.new
    @finder.find(@options)
  end

  def method_missing(method, *args, &block)
    engine = @@default_options[:supported_engines].find{ |name| ("from_" << name.to_s).match /#{method}/ }
    if engine
      @options.merge!({:engine => engine})
      send("from")
    else
      super(method, *args, &block)
    end
  end

  protected
  def validate_options
    raise "Keyword and site parameters must be Strings" unless @options[:keyword].is_a?(String) and @options[:url].is_a?(String)
    raise "Engine should be 'bing', 'google' or 'yahoo'" unless @@default_options[:supported_engines].include?(@options[:engine].to_sym)
  end

end

class Finder

  def find(options) #keyword, url, limit, engine, res_per_page
    options[:keyword].gsub!(/\s/, '+')
    request_url, results_selector, cite_selector = case options[:engine].to_sym
    when :bing
      ["http://www.bing.com/search?q=#{options[:keyword]}&count=#{options[:res_per_page]}&first=", '#wg0 > li', 'cite']
    when :google
      ["http://www.google.com/search?q=#{options[:keyword]}&num=#{options[:res_per_page]}&start=", '#ires > ol > li', 'cite']
    when :yahoo
      ["http://search.yahoo.com/search?p=#{options[:keyword]}&n=#{options[:res_per_page]}&b=", '#web > ol > li', 'span']
    end

    count, rank = 0, nil

    loop {
      html_response = Net::HTTP.get_response(URI.parse("#{request_url}#{count}")).body
      html_results = Nokogiri.parse(html_response).css(results_selector)
      rank = html_results.index(html_results.detect{ |result| result.css(cite_selector).text.match Regexp.new(options[:url]) })

      if count > options[:limit]
        break
      elsif rank
        rank += count
        break
      end
      count += options[:res_per_page]
    }

    rank ? rank.next : nil

   end

end



keyword_rankings/keyword_rankings.rb
# :offset => page number
# BING REAL LIMIT = 1000 pages of 10 results
# GOOGLE REAL LIMIT = 56 pages of ~ 63 results
# YAHOO REAL LIMIT = 100 results

module KeywordRankings
  RES_LIMIT = 200
  RES_PER_PAGE = 10 #careful, it doesn't work with more than 10 for all search engines!

  def get_ranking(*args) # keyword, url, engine, limit
    options = args.extract_options!
    limit = options[:limit].present? ? options[:limit].to_i : RES_LIMIT
    site_uri = options[:url].gsub(/(http:\/\/|https:\/\/)/, '').gsub(/^www/, '')
    validate_arguments options
    find_ranking(options[:keyword], site_uri, limit, options[:engine])
  end

  def validate_arguments(options)
    raise "Keyword and site parameters must be Strings" unless options[:keyword].present? and options[:keyword].is_a?(String) and options[:url].present? and options[:url].is_a?(String)
    raise "Limit of #{RES_LIMIT} results at most" if options[:limit] > RES_LIMIT
    raise "Engine should be 'bing', 'google' or 'yahoo'" unless options[:engine].present? and [:bing, :google, :yahoo].include?(options[:engine].to_sym)
  end

  def find_ranking(keyword, site, limit, engine)
    request_url, results_container, cite_container = case engine.to_sym
    when :bing
      ["http://www.bing.com/search?q=#{keyword}&count=#{RES_PER_PAGE}&first=", '#wg0 > li', 'cite']
    when :google
      ["http://www.google.com/search?q=#{keyword}&num=#{RES_PER_PAGE}&start=", '#ires > ol > li', 'cite']
    when :yahoo
      ["http://search.yahoo.com/search?p=#{keyword}&n=#{RES_PER_PAGE}&b=", '#web > ol > li', 'span']
    end
    count, rank = 0, nil
    loop {
      results = Nokogiri.parse(%x{ curl -A Mozilla "#{request_url}#{count}" }).css(results_container)
      rank = results.index(results.detect{ |result| result.css(cite_container).text.match Regexp.new(site) })
      if count > RES_LIMIT
        break
      elsif rank.present?
        rank = count + rank
        break
      end
      count += RES_PER_PAGE
    }
    rank.try(:next)
  end

end

